File: sdks/js_sdk/src/runtime/voting.ts
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Voting and aggregation strategies for SW4RM.
 *
 * This module provides aggregation strategies for combining critic votes
 * in negotiation room workflows. Strategies include:
 * - SimpleAverageAggregator: Arithmetic mean of all scores
 * - ConfidenceWeightedAggregator: Weight votes by confidence (POMDP-based)
 * - MajorityVoteAggregator: Count passed vs failed votes
 * - BordaCountAggregator: Ranked voting with position-based points
 *
 * Based on Python SDK voting/strategies.py and voting/aggregation.py.
 */

import {
  NegotiationVote,
  AggregatedScore,
  NegotiationValidationError,
} from '../clients/negotiationRoom.js';

/**
 * Interface for vote aggregation strategies.
 *
 * Implementations must provide an aggregate method that takes a list of votes
 * and returns an AggregatedScore with statistical summary.
 */
export interface VotingAggregator {
  /**
   * Aggregate votes into statistical summary.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with aggregation results
   * @throws NegotiationValidationError if votes list is empty or invalid
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore;
}

/**
 * Simple arithmetic mean aggregation strategy.
 *
 * Computes basic statistics without considering confidence levels.
 * All votes are weighted equally.
 */
export class SimpleAverageAggregator implements VotingAggregator {
  /**
   * Aggregate votes using simple arithmetic mean.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with simple mean in both mean and weightedMean fields
   * @throws NegotiationValidationError if votes list is empty
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot aggregate empty list of votes'
      );
    }

    const scores = votes.map((v) => v.score);

    // Basic statistics
    const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
    const minScore = Math.min(...scores);
    const maxScore = Math.max(...scores);

    // Standard deviation
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / scores.length;
    const stdDev = Math.sqrt(variance);

    // For simple average, weightedMean equals mean
    return {
      mean,
      minScore,
      maxScore,
      stdDev,
      weightedMean: mean,
      voteCount: votes.length,
    };
  }
}

/**
 * Confidence-weighted aggregation strategy based on POMDP research.
 *
 * Weights each vote's score by its confidence value, giving more influence
 * to votes from critics with higher certainty. This implements the approach
 * from POMDP research where agent confidence reflects belief state certainty.
 *
 * Formula: weighted_mean = sum(score_i * confidence_i) / sum(confidence_i)
 */
export class ConfidenceWeightedAggregator implements VotingAggregator {
  /**
   * Aggregate votes using confidence weighting.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with confidence-weighted mean
   * @throws NegotiationValidationError if votes list is empty
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot aggregate empty list of votes'
      );
    }

    const scores = votes.map((v) => v.score);
    const confidences = votes.map((v) => v.confidence);

    // Basic statistics (unweighted)
    const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
    const minScore = Math.min(...scores);
    const maxScore = Math.max(...scores);

    // Standard deviation (unweighted)
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / scores.length;
    const stdDev = Math.sqrt(variance);

    // Confidence-weighted mean
    const totalConfidence = confidences.reduce((a, b) => a + b, 0);
    let weightedMean: number;
    if (totalConfidence > 0) {
      weightedMean =
        scores.reduce((sum, s, i) => sum + s * confidences[i], 0) /
        totalConfidence;
    } else {
      // Fallback to simple mean if all confidences are zero
      weightedMean = mean;
    }

    return {
      mean,
      minScore,
      maxScore,
      stdDev,
      weightedMean,
      voteCount: votes.length,
    };
  }
}

/**
 * Majority voting aggregation strategy.
 *
 * Counts the number of passed vs failed votes and returns a score based on
 * the majority outcome. This is a binary approach that ignores the numerical
 * score values and focuses on the pass/fail decision.
 *
 * Score mapping:
 * - All passed: 10.0
 * - Majority passed (>50%): 7.5
 * - Tie: 5.0
 * - Majority failed (>50%): 2.5
 * - All failed: 0.0
 */
export class MajorityVoteAggregator implements VotingAggregator {
  /**
   * Aggregate votes using majority voting.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with score based on majority outcome
   * @throws NegotiationValidationError if votes list is empty
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot aggregate empty list of votes'
      );
    }

    const scores = votes.map((v) => v.score);
    const passedCount = votes.filter((v) => v.passed).length;
    const failedCount = votes.length - passedCount;

    // Determine majority outcome and assign score
    let majorityScore: number;
    if (passedCount === votes.length) {
      // All passed
      majorityScore = 10.0;
    } else if (failedCount === votes.length) {
      // All failed
      majorityScore = 0.0;
    } else if (passedCount > failedCount) {
      // Majority passed
      majorityScore = 7.5;
    } else if (failedCount > passedCount) {
      // Majority failed
      majorityScore = 2.5;
    } else {
      // Tie
      majorityScore = 5.0;
    }

    // Basic statistics on original scores
    const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
    const minScore = Math.min(...scores);
    const maxScore = Math.max(...scores);
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / scores.length;
    const stdDev = Math.sqrt(variance);

    return {
      mean,
      minScore,
      maxScore,
      stdDev,
      weightedMean: majorityScore,
      voteCount: votes.length,
    };
  }
}

/**
 * Borda count aggregation strategy.
 *
 * Implements ranked voting where each vote receives points based on its
 * rank position. Higher scored votes receive more points.
 *
 * Ranking system:
 * - Votes are sorted by score (highest to lowest)
 * - Highest score receives n points (where n = number of votes)
 * - Second highest receives n-1 points
 * - Lowest score receives 1 point
 * - Points are summed and normalized to 0-10 scale
 *
 * This method is resistant to strategic voting and reduces impact of outliers.
 */
export class BordaCountAggregator implements VotingAggregator {
  /**
   * Aggregate votes using Borda count.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with Borda count normalized score
   * @throws NegotiationValidationError if votes list is empty
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot aggregate empty list of votes'
      );
    }

    const scores = votes.map((v) => v.score);
    const n = votes.length;

    // Basic statistics
    const mean = scores.reduce((a, b) => a + b, 0) / n;
    const minScore = Math.min(...scores);
    const maxScore = Math.max(...scores);
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / n;
    const stdDev = Math.sqrt(variance);

    // Borda count calculation
    // Sort scores with their indices to handle ties
    const indexedScores = scores.map((score, idx) => ({ score, idx }));
    indexedScores.sort((a, b) => b.score - a.score);

    // Assign points based on rank (highest gets n points, lowest gets 1)
    let totalPoints = 0;
    for (let rank = 0; rank < indexedScores.length; rank++) {
      const points = n - rank;
      totalPoints += points;
    }

    // Normalize to 0-10 scale
    // Average points per position
    const avgPoints = totalPoints / n;
    // Normalize: avgPoints / n * 10 gives us the normalized score
    const bordaScore = (avgPoints / n) * 10.0;

    return {
      mean,
      minScore,
      maxScore,
      stdDev,
      weightedMean: bordaScore,
      voteCount: n,
    };
  }
}

/**
 * Main aggregator class with consensus and polarization detection.
 *
 * Wraps a VotingAggregator strategy and provides additional methods for
 * analyzing vote distributions, detecting consensus, and identifying
 * polarization.
 */
export class VotingAnalyzer {
  private strategy: VotingAggregator;

  /**
   * Initialize analyzer with a strategy.
   *
   * @param strategy - The aggregation strategy to use
   */
  constructor(strategy: VotingAggregator) {
    this.strategy = strategy;
  }

  /**
   * Aggregate votes using the configured strategy.
   *
   * @param votes - List of critic votes to aggregate
   * @returns AggregatedScore with aggregation results
   * @throws NegotiationValidationError if votes list is empty
   */
  aggregate(votes: NegotiationVote[]): AggregatedScore {
    return this.strategy.aggregate(votes);
  }

  /**
   * Compute Shannon entropy of vote distribution.
   *
   * Entropy measures the uncertainty or disorder in the vote distribution.
   * Higher entropy indicates more disagreement/uncertainty, lower entropy
   * indicates more consensus.
   *
   * The score range [0, 10] is divided into bins, and entropy is computed
   * over the probability distribution of votes across bins.
   *
   * @param votes - List of critic votes
   * @returns Entropy value in bits (0 = perfect consensus, higher = more uncertainty)
   * @throws NegotiationValidationError if votes list is empty
   */
  computeEntropy(votes: NegotiationVote[]): number {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot compute entropy of empty vote list'
      );
    }

    // Create bins for score distribution (0-2, 2-4, 4-6, 6-8, 8-10)
    const numBins = 5;
    const binWidth = 10.0 / numBins;
    const binCounts = new Array(numBins).fill(0);

    // Count votes in each bin
    for (const vote of votes) {
      const binIdx = Math.min(Math.floor(vote.score / binWidth), numBins - 1);
      binCounts[binIdx]++;
    }

    // Compute Shannon entropy
    const total = votes.length;
    let entropy = 0.0;
    for (const count of binCounts) {
      if (count > 0) {
        const probability = count / total;
        entropy -= probability * Math.log2(probability);
      }
    }

    return entropy;
  }

  /**
   * Detect if votes show strong consensus.
   *
   * Consensus is detected when:
   * 1. Standard deviation of scores is low (< 2.0)
   * 2. A large proportion of critics agree on pass/fail (>= threshold)
   *
   * @param votes - List of critic votes
   * @param threshold - Minimum proportion of agreement required (default 0.8 = 80%)
   * @returns True if consensus is detected, false otherwise
   * @throws NegotiationValidationError if votes list is empty or threshold not in [0, 1]
   */
  detectConsensus(votes: NegotiationVote[], threshold: number = 0.8): boolean {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot detect consensus in empty vote list'
      );
    }
    if (threshold < 0.0 || threshold > 1.0) {
      throw new NegotiationValidationError(
        `Threshold must be in [0, 1], got ${threshold}`
      );
    }

    // Compute standard deviation
    const scores = votes.map((v) => v.score);
    const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / scores.length;
    const stdDev = Math.sqrt(variance);

    // Check if scores are tightly clustered
    if (stdDev >= 2.0) {
      return false;
    }

    // Check if majority agrees on pass/fail
    const passedCount = votes.filter((v) => v.passed).length;
    const failedCount = votes.length - passedCount;

    const agreementRatio = Math.max(passedCount, failedCount) / votes.length;
    return agreementRatio >= threshold;
  }

  /**
   * Detect if votes show polarization (bimodal distribution).
   *
   * Polarization is detected when votes cluster into two distinct groups,
   * typically at opposite ends of the score spectrum. This is identified by:
   * 1. High standard deviation (>= 3.0)
   * 2. Low density in the middle range (4-6)
   * 3. High density in the extremes (0-3 or 7-10)
   *
   * @param votes - List of critic votes
   * @returns True if polarization is detected, false otherwise
   * @throws NegotiationValidationError if votes list is empty
   */
  detectPolarization(votes: NegotiationVote[]): boolean {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot detect polarization in empty vote list'
      );
    }

    if (votes.length < 3) {
      // Need at least 3 votes to meaningfully detect polarization
      return false;
    }

    // Compute standard deviation
    const scores = votes.map((v) => v.score);
    const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
    const variance =
      scores.reduce((sum, s) => sum + Math.pow(s - mean, 2), 0) / scores.length;
    const stdDev = Math.sqrt(variance);

    // High standard deviation suggests spread
    if (stdDev < 3.0) {
      return false;
    }

    // Count votes in three regions: low (0-3), middle (4-6), high (7-10)
    const lowCount = scores.filter((s) => s <= 3.0).length;
    const middleCount = scores.filter((s) => s > 3.0 && s < 7.0).length;
    const highCount = scores.filter((s) => s >= 7.0).length;

    // Polarization: both extremes have votes, middle is sparse
    const hasLowCluster = lowCount >= votes.length * 0.3;
    const hasHighCluster = highCount >= votes.length * 0.3;
    const middleSparse = middleCount <= votes.length * 0.2;

    return hasLowCluster && hasHighCluster && middleSparse;
  }

  /**
   * Compute variance of confidence levels across votes.
   *
   * High variance in confidence suggests critics have different levels of
   * certainty about their evaluations, which may indicate the artifact is
   * harder to assess or has ambiguous quality.
   *
   * @param votes - List of critic votes
   * @returns Variance of confidence values
   * @throws NegotiationValidationError if votes list is empty
   */
  computeConfidenceVariance(votes: NegotiationVote[]): number {
    if (votes.length === 0) {
      throw new NegotiationValidationError(
        'Cannot compute confidence variance of empty vote list'
      );
    }

    const confidences = votes.map((v) => v.confidence);
    const meanConfidence =
      confidences.reduce((a, b) => a + b, 0) / confidences.length;
    const variance =
      confidences.reduce(
        (sum, c) => sum + Math.pow(c - meanConfidence, 2),
        0
      ) / confidences.length;

    return variance;
  }

  /**
   * Get comprehensive summary of vote characteristics.
   *
   * Combines multiple analytical methods to provide a complete picture
   * of the vote distribution and critic agreement.
   *
   * @param votes - List of critic votes
   * @returns Object with summary statistics
   * @throws NegotiationValidationError if votes list is empty
   */
  getVoteSummary(votes: NegotiationVote[]): {
    entropy: number;
    consensus: boolean;
    polarization: boolean;
    confidenceVariance: number;
    passRate: number;
  } {
    if (votes.length === 0) {
      throw new NegotiationValidationError('Cannot summarize empty vote list');
    }

    const passedCount = votes.filter((v) => v.passed).length;

    return {
      entropy: this.computeEntropy(votes),
      consensus: this.detectConsensus(votes),
      polarization: this.detectPolarization(votes),
      confidenceVariance: this.computeConfidenceVariance(votes),
      passRate: passedCount / votes.length,
    };
  }
}

File: sdks/rust_sdk/examples/voting.rs
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Voting aggregation example for the SW4RM Rust SDK.
//!
//! This example demonstrates how to use different aggregation strategies to
//! combine critic votes in a negotiation room scenario. It mirrors the
//! Python SDK example at `sdks/py_sdk/examples/voting_example.py`.
//!
//! Four aggregation strategies are demonstrated:
//!
//! - `SimpleAverageAggregator`: Arithmetic mean of all scores
//! - `ConfidenceWeightedAggregator`: Weights each vote by its confidence level
//! - `MajorityVoteAggregator`: Binary pass/fail majority counting
//! - `BordaCountAggregator`: Ranked voting with position-based points
//!
//! Additional analytical capabilities shown:
//!
//! - Shannon entropy of vote distribution
//! - Consensus detection (low std dev + high pass/fail agreement)
//! - Polarization detection (bimodal score distribution)
//! - Confidence variance across critics
//!
//! Run with:
//!
//! ```bash
//! cargo run --example voting
//! ```

use anyhow::Result;
use sw4rm_sdk::clients::negotiation_room::NegotiationVote;
use sw4rm_sdk::{
    BordaCountAggregator, ConfidenceWeightedAggregator, MajorityVoteAggregator,
    SimpleAverageAggregator, VotingAggregator,
};

/// Create a realistic set of critic votes for demonstration.
///
/// Four critics evaluate the same artifact from different perspectives:
/// security, performance, code quality, and testing coverage.
fn create_example_votes() -> Result<Vec<NegotiationVote>> {
    let votes = vec![
        NegotiationVote::new(
            "artifact_123".to_string(),
            "security_critic".to_string(),
            8.5,
            0.9,
            true,
            vec![
                "Strong authentication".to_string(),
                "Good input validation".to_string(),
            ],
            vec!["Missing rate limiting".to_string()],
            vec!["Add rate limiting to API endpoints".to_string()],
            "room_001".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_123".to_string(),
            "performance_critic".to_string(),
            7.0,
            0.75,
            true,
            vec!["Efficient database queries".to_string()],
            vec!["No caching strategy".to_string()],
            vec!["Consider adding Redis cache".to_string()],
            "room_001".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_123".to_string(),
            "code_quality_critic".to_string(),
            6.5,
            0.8,
            true,
            vec![
                "Clean code structure".to_string(),
                "Good documentation".to_string(),
            ],
            vec!["Some functions are too long".to_string()],
            vec!["Refactor large functions into smaller ones".to_string()],
            "room_001".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_123".to_string(),
            "testing_critic".to_string(),
            9.0,
            0.95,
            true,
            vec![
                "Excellent test coverage".to_string(),
                "Good edge case handling".to_string(),
            ],
            vec![],
            vec!["Add more integration tests".to_string()],
            "room_001".to_string(),
        )?,
    ];

    Ok(votes)
}

/// Demonstrate all four aggregation strategies with the example votes.
///
/// Prints a formatted comparison of each strategy's output, followed by
/// analytical metrics (entropy, consensus, polarization, confidence variance)
/// and a comprehensive summary.
fn demonstrate_strategies() -> Result<()> {
    let votes = create_example_votes()?;

    println!("{}", "=".repeat(70));
    println!("Voting Aggregation Strategies Demonstration");
    println!("{}", "=".repeat(70));
    println!("\nNumber of votes: {}", votes.len());
    println!("\nIndividual votes:");
    for vote in &votes {
        println!(
            "  {:<25}: score={:.1}, confidence={:.2}, passed={}",
            vote.critic_id, vote.score, vote.confidence, vote.passed
        );
    }

    println!("\n{}", "=".repeat(70));
    println!("Strategy Comparison");
    println!("{}", "=".repeat(70));

    // 1. Simple Average
    println!("\n1. Simple Average Strategy:");
    println!("   Treats all votes equally");
    let simple_agg = VotingAggregator::new(Box::new(SimpleAverageAggregator::new()));
    let simple_result = simple_agg.aggregate(&votes)?;
    println!("   Mean score:    {:.2}", simple_result.mean);
    println!("   Weighted mean: {:.2}", simple_result.weighted_mean);
    println!("   Std deviation: {:.2}", simple_result.std_dev);

    // 2. Confidence Weighted
    println!("\n2. Confidence-Weighted Strategy:");
    println!("   Weights votes by critic confidence (POMDP-based)");
    let conf_agg = VotingAggregator::new(Box::new(ConfidenceWeightedAggregator::new()));
    let conf_result = conf_agg.aggregate(&votes)?;
    println!("   Mean score:    {:.2}", conf_result.mean);
    println!("   Weighted mean: {:.2}", conf_result.weighted_mean);
    println!("   -> Testing critic (0.95 conf, 9.0 score) has more influence");

    // 3. Majority Vote
    println!("\n3. Majority Vote Strategy:");
    println!("   Based on pass/fail count (ignores numerical scores)");
    let majority_agg = VotingAggregator::new(Box::new(MajorityVoteAggregator::new()));
    let majority_result = majority_agg.aggregate(&votes)?;
    println!("   Mean score:      {:.2}", majority_result.mean);
    println!("   Majority outcome: {:.2}", majority_result.weighted_mean);
    println!("   -> All critics passed, so score is 10.0");

    // 4. Borda Count
    println!("\n4. Borda Count Strategy:");
    println!("   Ranked voting with position-based points");
    let borda_agg = VotingAggregator::new(Box::new(BordaCountAggregator::new()));
    let borda_result = borda_agg.aggregate(&votes)?;
    println!("   Mean score:  {:.2}", borda_result.mean);
    println!("   Borda score: {:.2}", borda_result.weighted_mean);
    println!("   -> Ranks: testing(9.0)->security(8.5)->performance(7.0)->quality(6.5)");

    // Analytical methods (reuse conf_agg for analysis)
    println!("\n{}", "=".repeat(70));
    println!("Vote Analysis");
    println!("{}", "=".repeat(70));

    let entropy = conf_agg.compute_entropy(&votes)?;
    println!("\nEntropy: {:.3} bits", entropy);
    println!("  (Low entropy = high agreement, High entropy = high disagreement)");

    let consensus = conf_agg.detect_consensus(&votes, 0.8)?;
    println!("\nConsensus detected: {}", consensus);
    println!("  (Low std dev + high pass/fail agreement)");

    let polarization = conf_agg.detect_polarization(&votes)?;
    println!("\nPolarization detected: {}", polarization);
    println!("  (Bimodal distribution with votes at extremes)");

    let conf_variance = conf_agg.compute_confidence_variance(&votes)?;
    println!("\nConfidence variance: {:.4}", conf_variance);
    println!("  (How much critics' confidence levels vary)");

    // Comprehensive summary
    println!("\n{}", "=".repeat(70));
    println!("Comprehensive Summary");
    println!("{}", "=".repeat(70));
    let summary = conf_agg.get_vote_summary(&votes)?;
    println!("\nPass rate:            {:.1}%", summary.pass_rate * 100.0);
    println!("Entropy:              {:.3} bits", summary.entropy);
    println!(
        "Consensus:            {}",
        if summary.consensus { "Yes" } else { "No" }
    );
    println!(
        "Polarization:         {}",
        if summary.polarization { "Yes" } else { "No" }
    );
    println!("Confidence variance:  {:.4}", summary.confidence_variance);

    println!("\n{}", "=".repeat(70));

    Ok(())
}

/// Demonstrate polarization detection with a strongly divided vote set.
///
/// Two critics give high scores (8.5-9.0) while two others give very low
/// scores (1.5-2.0), creating a bimodal distribution that triggers the
/// polarization detector. This scenario typically warrants HITL escalation.
fn demonstrate_polarization() -> Result<()> {
    println!("\n{}", "=".repeat(70));
    println!("Polarization Example");
    println!("{}", "=".repeat(70));

    let polarized_votes = vec![
        NegotiationVote::new(
            "artifact_456".to_string(),
            "critic_1".to_string(),
            9.0,
            0.85,
            true,
            vec!["Excellent".to_string()],
            vec![],
            vec![],
            "room_002".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_456".to_string(),
            "critic_2".to_string(),
            8.5,
            0.9,
            true,
            vec!["Very good".to_string()],
            vec![],
            vec![],
            "room_002".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_456".to_string(),
            "critic_3".to_string(),
            2.0,
            0.8,
            false,
            vec![],
            vec!["Major issues".to_string()],
            vec!["Complete rewrite".to_string()],
            "room_002".to_string(),
        )?,
        NegotiationVote::new(
            "artifact_456".to_string(),
            "critic_4".to_string(),
            1.5,
            0.75,
            false,
            vec![],
            vec!["Critical flaws".to_string()],
            vec!["Start over".to_string()],
            "room_002".to_string(),
        )?,
    ];

    println!("\nPolarized votes (critics strongly disagree):");
    for vote in &polarized_votes {
        println!(
            "  {}: score={:.1}, passed={}",
            vote.critic_id, vote.score, vote.passed
        );
    }

    let aggregator = VotingAggregator::new(Box::new(ConfidenceWeightedAggregator::new()));
    let result = aggregator.aggregate(&polarized_votes)?;

    println!("\nSimple mean:   {:.2}", result.mean);
    println!("Std deviation: {:.2} (high variance)", result.std_dev);
    println!(
        "Entropy:       {:.3} bits",
        aggregator.compute_entropy(&polarized_votes)?
    );
    println!(
        "Consensus:     {}",
        aggregator.detect_consensus(&polarized_votes, 0.8)?
    );
    println!(
        "Polarization:  {} <- Detected!",
        aggregator.detect_polarization(&polarized_votes)?
    );

    println!("\n-> This indicates the artifact needs human review (HITL escalation)");

    Ok(())
}

fn main() -> Result<()> {
    demonstrate_strategies()?;
    demonstrate_polarization()?;

    println!("\n{}", "=".repeat(70));
    println!("Example completed successfully!");
    println!("{}", "=".repeat(70));

    Ok(())
}
